### PR TITLE
[Merged by Bors] - feat(CategoryTheory): the left Kan extension functor

### PR DESCRIPTION
**Pull request by joelriou**
Given a functor `F : C ⥤ D`, we define the left Kan extension functor `F.lan : (C ⥤ E) ⥤ (D ⥤ E)` which sends a functor `G : C ⥤ E` to its left Kan extension along `F`. (This is a step towards the refactor of `Lan/Ran` in mathlib using the new API for Kan extensions of functors #10425.)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Functor/KanExtension/Adjunction.lean [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+variable (E) in
+@[simp]
+lemma lanAdjunction_unit :
+    (F.lanAdjunction E).unit =
+      lanUnit F := by
+  ext G : 2
+  dsimp [lanAdjunction, homEquivOfIsLeftKanExtension]
+  simp
+
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Functor/KanExtension/Adjunction.lean:83]:
```suggestion
@[simp]
lemma lanAdjunction_counit_app (G : D ⥤ E) : (F.lanAdjunction E).counit.app G = 
    descOfIsLeftKanExtension (F.lan.obj (F ⋙ G)) (F.lanUnit.app (F ⋙ G)) G (𝟙 (F ⋙ G)) := rfl
```
I don't know if this is necessary but it seems desireable to have access to the API lemmas for `descOfIsLeftKanExtension` when working with the counit

**Review by dagurtomas — COMMENTED** [review]
Could you make the variable names throughout the folder `Functor/KanExtension` more consistent? For example, always use `L: C ⥤ D` for the "along" functor and `F : C ⥤ E` for the "of" functor (or whatever). It's a bit confusing that `F` becomes the "along" functor in the adjunction file when it was the "of" functor in the basic file, and also the categories changed names.

File: Mathlib/CategoryTheory/Functor/KanExtension/Adjunction.lean (hunk 1 above)
Comment (joelriou) [reply Mathlib/CategoryTheory/Functor/KanExtension/Adjunction.lean:89]: Thanks for the suggestion. I have added this lemma, but not as a `simp` lemma in https://github.com/leanprover-community/mathlib4/pull/12168/commits/1ce90141d041fc31386752eb402e52dbbe7a2a0b where I also obtained the lemma
```lean
lemma isIso_lanAdjunction_counit_app_iff (G : D ⥤ E) :
    IsIso ((F.lanAdjunction E).counit.app G) ↔ G.IsLeftKanExtension (𝟙 (F ⋙ G)) :=
```
Here I feel that it is better if the LHS is not further simplified.

**Comment by joelriou** [comment]
Thanks @dagurtomas for the suggestion. It was indeed very confusing... Now, we extend `F : C ⥤ H` along `L : C ⥤ D`. I use the category `H` because `E` is sometimes used for terms in `LeftExtension L F`. (Doing so, the notations will be more consistent with the future API for derived functors, where `L` shall be a localization functor.)

**Review by dagurtomas — APPROVED** [review]
Thanks! LGTM

**Comment by erdOne** [comment]
Thanks!
maintainer merge

File: Mathlib/CategoryTheory/Functor/KanExtension/Adjunction.lean [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+Given a functor `L : C ⥤ D`, we define the left Kan extension functor
+`L.lan : (C ⥤ H) ⥤ (D ⥤ H)` which sends a functor `F : C ⥤ H` to its
+left Kan extension along `L`. This is defined if all `F` have such
+a left Kan extension. It is shown that `L.lan` is the left adjoint to
+the functor `(D ⥤ H) ⥤ (C ⥤ H)` given by the precomposition
+with `L` (see `Functor.lanAdjunction`).
+
+## TODO
+- dualize the results for right Kan extensions
+- refactor the file `CategoryTheory.Limits.KanExtension`
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Functor/KanExtension/Adjunction.lean:19]: Can you please add 1 or 2 sentences explaining what this refactor should be doing?

**Review by jcommelin — APPROVED** [review]
Thanks so much! 1 minor request.

bors d+

File: Mathlib/CategoryTheory/Functor/KanExtension/Adjunction.lean (hunk 2 above)
Comment (joelriou) [reply Mathlib/CategoryTheory/Functor/KanExtension/Adjunction.lean]: Thanks for the suggestion. I have added more details.

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**